add SQS feature flag to raise QueueDeletedRecently errors

thrau opened the pull request:
This PR adds a feature flag for enabling the behavior requested in
-  #5801

AWS does not allow creating a queue with the same name for 60 seconds after it was deleted. See the [`DeleteQueue` API Reference](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_DeleteQueue.html). By starting localstack with `SQS_DELAY_RECENTLY_DELETED=1`, localstack will behave in the same way, but it is turned off by default.

The reason for adding the feature flag was that, localstack is often used in testing scenarios where test suites may depend on having a queue with the same name be deleted and created frequently.

viren-nadkarni reviewed (approved):
Looks good :+1: 

Just need to update user docs

bentsku commented on localstack/services/sqs/provider.py:
```diff
@@ -795,6 +798,19 @@ def create_queue(
+
+        # When you delete a queue, you must wait at least 60 seconds before creating a queue with the same name.
+        # see https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_DeleteQueue.html
+        if queue_name in backend.deleted and config.SQS_DELAY_RECENTLY_DELETED:
+            deleted = backend.deleted[queue_name]
+            if deleted > (time.time() - 60):
+                raise QueueDeletedRecently(
+                    "You must wait 60 seconds after deleting a queue before you can create another with the same name."
+                )
+            else:
+                del backend.deleted[queue_name]
```
I am not sure about the lifecycle of the provider, but I think we are only deleting the entry in `self.deleted` if `config.SQS_DELAY_RECENTLY_DELETED` is `True`. But we are always adding the creation time when deleting in `delete_queue`. Not sure if important, but it could lead to memory leaks? We're using only the queue name as as key but the queue reference could be kept.

(Also, this is maybe my personal preference and I have no idea if it does change anything or if it's optimised internally afterwards, but I put the boolean check before the dictionary lookup to avoid it if the flag is not set. But I don't know if it's useful or not.)

thrau replied:
both great points :-)

bentsku reviewed (changes requested):
(Sorry, I'm fighting with my own life against GitHub UI for requesting changes, hope it doesn't make a mess 😅)

thrau commented:
updated the PR @bentsku according to your suggestions
* added an `expire_deleted` method that's called every time in the `create_queue` method to clear up the cache.
* parameterized the 60 second timeout to be able to write a whitebox test
* moved the entire `create_queue` logic (which is a critical section) into the `mutex` scope to protect it from concurrency issues.

bentsku commented:
> updated the PR @bentsku according to your suggestions:
> 
> * added an `expire_deleted` method that's called every time in the `create_queue` method to clear up the cache
> 
> and also:
> 
> * parameterized the 60 second timeout to be able to write a whitebox test
> * moved the entire `create_queue` logic (which is a critical section) into the `mutex` scope to protect it from concurrency issues.

I actually didn't think at all about the cleanup, this is really good!
I think we should call the cleanup even if the flag isn't set though, because we always create a new entry in `self.deleted` when deleting a queue. I think just outdent `self.expire_deleted` to the left to be outside the flag conditional will do.

bentsku commented on localstack/services/sqs/provider.py:
```diff
@@ -791,18 +803,31 @@ def create_queue(
+                        "You must wait 60 seconds after deleting a queue before you can create "
+                        "another with the same name."
+                    )
+                backend.expire_deleted()
```
I'd outdent this one to be called outside the `if config.SQS_DELAY_RECENTLY_DELETED:`, so it would cleanup even when the flag isn't set.
